### PR TITLE
[Storage] Storing and indexing multiple receipts for same block

### PR DESCRIPTION
**Pull request by durkmurder**
This PR implements support for storing multiple receipts for same block. Introduces separation between execution receipts storage for execution and non-execution nodes.
Key points:
- Indexing happens when storing as one atomic operation.
- Execution node indexes only one receipt for block, non-execution nodes multiple.
- Removed usage of all manual indexing and indexers.

### CODE REVIEW

File: storage/badger/receipts.go [hunk 1]
```diff
@@ -99,9 +91,17 @@ func (r *ExecutionReceipts) ByID(receiptID flow.Identifier) (*flow.ExecutionRece
...
+	tx := r.db.NewTransaction(false)
+	defer tx.Discard()
+	return r.byBlockId(tx, blockID)
+}
 
+func NewExecutionReceipts(collector module.CacheMetrics, db *badger.DB, results *ExecutionResults) *ExecutionReceipts {
+	r := newExecutionReceipts(collector, db, results)
+	r.index = func(tx *badger.Txn, receipt *flow.ExecutionReceipt) error {
+		blockID := receipt.ExecutionResult.BlockID
+		receiptID := receipt.ID()
```
Comment (durkmurder) [inline storage/badger/receipts.go:104]: Still failing when we are trying to index different result. For fork recovery it probably makes sense to update entry if it's already there. Should I do this?

File: storage/badger/receipts.go [hunk 2]
```diff
@@ -140,16 +140,29 @@ func (r *ExecutionReceipts) Index(blockID, receiptID flow.Identifier) error {
...
+	}
+	return r
 }
 
-func (r *ExecutionReceipts) IndexByExecutor(receipt *flow.ExecutionReceipt) error {
-	blockID := receipt.ExecutionResult.BlockID
-	executorID := receipt.ExecutorID
-	receiptID := receipt.ID()
-	return operation.RetryOnConflict(r.db.Update, func(tx *badger.Txn) error {
+func NewAllExecutionReceipts(collector module.CacheMetrics, db *badger.DB, results *ExecutionResults) *ExecutionReceipts {
```
Comment (yhassanzadeh13) [inline storage/badger/receipts.go:159]: The demarcation between `NewExecutionReceipts` and `NewAllExecutionReceipts` is very tiny and may lead to confusion. Maybe renaming to `NewNonExecutionNodeReceipts` and `NewExecutionNodeReceipts` would help a more conveying name. The other option is to divide the interface `ExecutionReceipts` into `ExecutionNodeReceipts` and `NonExecutionNodeReceipts`. This is though just a suggestion for readability and maintenance. I leave it totally up to you. Definitely not a blocker.

File: storage/badger/receipts.go [hunk 3]
```diff
@@ -140,16 +140,29 @@ func (r *ExecutionReceipts) Index(blockID, receiptID flow.Identifier) error {
...
+		var receiptID flow.Identifier
+		err := operation.LookupExecutionReceipt(blockID, &receiptID)(tx)
+		if err != nil {
+			return nil, fmt.Errorf("could not lookup receipt ID: %w", err)
+		}
+		receipt, err := r.byID(receiptID)(tx)
+		if err != nil {
+			return nil, err
+		}
+		return []*flow.ExecutionReceipt{receipt}, nil
```
Comment (yhassanzadeh13) [inline storage/badger/receipts.go:154]: From [this part](https://github.com/onflow/flow-go/pull/486/files#diff-d8547dad4f02ecbdc866e2778645382aa5cd89ba25a5f9441cc0e0c094218429R36-R43) it seems that possible to return an _empty_ slice even when our error is `nil`. If that is the case, I would do the check here and return a _not found error_, so that we do not need any extra checking externally other than checking the error itself.

**Review by yhassanzadeh13 — APPROVED** [review]
Overall looks good, though I am not very familiar with the domain. Just a few minor comments for readability.

File: storage/badger/operation/receipts.go [hunk 4]
```diff
@@ -18,39 +18,42 @@ func RetrieveExecutionReceiptMeta(receiptID flow.Identifier, meta *flow.Executio
...
 		check := func(key []byte) bool {
-			return true
+			if len(key) == len(val)*2+1 {
+				offset := len(val) + 1
+				copy(val[:], key[offset:])
+				return true
+			} else {
+				return false
+			}
 		}
```
Comment (zhangchiqing) [inline storage/badger/operation/receipts.go:52]: can you explain how this works?

File: storage/badger/receipts.go (hunk 2 above)
Comment (zhangchiqing) [inline storage/badger/receipts.go:159]: Better to add comments and highlight the difference between it and NewExecutionReceipt

**Review by zhangchiqing — COMMENTED** [review]
Overall looks great. Had some comments and questions

File: storage/badger/operation/receipts.go (hunk 4 above)
Comment (durkmurder) [reply storage/badger/operation/receipts.go:52]: kinda dark magic but not sure if there is another way.
We build our prefix like this:
```
func makePrefix(code byte, keys ...interface{}) []byte {
	prefix := make([]byte, 1)
	prefix[0] = code
	for _, key := range keys {
		prefix = append(prefix, b(key)...)
	}
	return prefix
}
```

it's basically concatenation of bytes. Since we are indexing by `blockID` and `receiptID` our pair is:
```
insert(makePrefix(codeAllBlockReceipt, blockID, receiptID), struct{}{})
```

When traversing badger db, I am checking for the first part of prefix: 
```
traverse(makePrefix(codeAllBlockReceipt, blockID), iterationFunc)
```
and when I have a partial match I extract the receiptID from it.
Hacky, I know, but it's the only way as I know with nosql databases.

File: storage/badger/receipts.go (hunk 1 above)
Comment (durkmurder) [reply storage/badger/receipts.go:104]: 1. I've just took what there was already implemented.
2. Completely makes sense. Currently `ExecutionReceipts` and `AllExecutionReceipts` uses different indexes so you can handle it easily.

File: storage/badger/receipts.go (hunk 2 above)
Comment (durkmurder) [reply storage/badger/receipts.go:159]: Agree, I was thinking something like:
`ExecutionReceiptsSingleIndex` `ExecutionReceiptsMutiIndex`
or 
`ExecutionNodeReceipts` `ConsensuNodeReceipts` (even though it's used by all nodes except execution). 
open to suggestions :)

File: storage/badger/receipts.go (hunk 3 above)
Comment (durkmurder) [reply storage/badger/receipts.go:154]: Basically did that to ensure that it's not nil. I am not sure how should we behave.
Suppose we don't have any receipts for `blockA`, should we return `([], nil)` or `(nil, err.NotFound)`.
That's an open question.

File: storage/badger/receipts.go [hunk 5]
```diff
@@ -140,16 +140,29 @@ func (r *ExecutionReceipts) Index(blockID, receiptID flow.Identifier) error {
...
 
 		return nil
-	})
+	}
+	r.byBlockId = func(tx *badger.Txn, blockID flow.Identifier) ([]*flow.ExecutionReceipt, error) {
+		var receiptID flow.Identifier
+		err := operation.LookupExecutionReceipt(blockID, &receiptID)(tx)
+		if err != nil {
+			return nil, fmt.Errorf("could not lookup receipt ID: %w", err)
+		}
```
[the remark this answers is not in the document]
Comment (durkmurder) [reply storage/badger/receipts.go:149]: Yeah, as I said, really not sure about it. See my [comment](https://github.com/onflow/flow-go/pull/486#discussion_r587822748). I don't have any preference so can do either way.

File: storage/badger/receipts.go (hunk 3 above)
Comment (durkmurder) [reply storage/badger/receipts.go:154]: - When returning error then you can be sure that content is nil and you can't use it, can't do smth like `receipts[0]`. But when error is nil you can be safe to get first receipt without checking for `len()`.
- When there is no error you need to check if slice is not empty before accessing it.

**Comment by AlexHentschel** [comment]
Thanks for putting up the PR. Your implementation quite closely follows my suggestion in [#5326](https://github.com/dapperlabs/flow-go/issues/5326). I am really sorry, reviewing your code made me discover a design flaw in my proposal 🤦 :

* Execution nodes (as any other nodes) follow the chain using the HotStuff follower. 
* The HotStuff follower persists blocks in storage:
   * via [`Blocks`](https://github.com/onflow/flow-go/blob/master/storage/badger/blocks.go), 
   * which then recursively stores the payload including the incorporated receipts 
   * at that point, many receipts from all ENs will hit the storage layer
   * meaning, if we are indexing receipts atomically on `Store`, we should be indexing all receipts
 *  The underlying design problem is that the https://github.com/onflow/flow-go/blob/master/storage/receipts.go Interface is used by generic components that all nodes share (like the HotStuff follower). I think if there are APIs that are used by generic components, the behaviour of the respective implementations should behave uniformly across all nodes. I only realized now that this is not the case:
    * The `ByBlockID` behaves fundamentally different for ENs
    * Also, I am worried that when the HotStuff follower is storing a block on the Execution Node, it might error, if there are conflicting receipts in the payload. 
 
As a result, I would like to suggest the following:
* The [`ExecutionReceipts`](https://github.com/onflow/flow-go/blob/82113bba53e963f96dd235ce15cd910dcd73874c/storage/receipts.go) API should have the same behaviour across 
all nodes.
* Only the execution node has the notion of "my execution receipt". But it still should understand that receipts exist from other nodes, which might be conflicting. Therefore, I would propose to make an extended Interface of `ExecutionReceipts`, e.g. [`MyExecutionReceipts`](https://github.com/onflow/flow-go/blob/e473b3819550f55ee2943d89a3662ed455b8205a/storage/receipts.go#L26-L38), which is only useful for Execution Nodes. 
* For the implementation, I would construct `MyExecutionReceipts` as a wrapper around `ExecutionReceipts`. The would be consistent with the implementation pattern of all the other storage-layer abstractions, for example:
   * `ExecutionReceipts` itself is a [wrapper around `ExecutionResults`](https://github.com/onflow/flow-go/blob/abdbdf4706d86b97697c44b6a92ecd4f017dc533/storage/badger/receipts.go#L19)
   * `Payloads` itself is a wrapper around, `ExecutionReceipts`, `Seals` and `Guarantees`

I have implemented my suggestions in this PR https://github.com/onflow/flow-go/pull/491 (targeting your branch). There are probably still some minor trickle-down effects to be fixed (I expect a few tests would are failing). Please let me know what you think.

File: storage/badger/operation/receipts.go (hunk 4 above)
Comment (AlexHentschel) [reply storage/badger/operation/receipts.go:52]: Thanks for the explanation @durkmurder . From my perspective, we can take two possible implementation approaches:

#### option 1 favours performance (Yurii' approach)
Yurii's approach has minimal memory footprint in the DB. If we are favouring performance, I think we should apply a few more optimizations:
* All the information we are interested in is already contained in the key, and all necessary processing steps can be done in the `check` method
   * In `check`, you already extract the relevant bytes from the key, why not append them to `receiptIDs` right away?
   * For each key, we decode the value: https://github.com/onflow/flow-go/blob/82113bba53e963f96dd235ce15cd910dcd73874c/storage/badger/operation/common.go#L293-L310 which is entirely unnecessary. If the `check` function returns false, [`traverse` would skip](https://github.com/onflow/flow-go/blob/82113bba53e963f96dd235ce15cd910dcd73874c/storage/badger/operation/common.go#L350-L353) all this unnecessary work.
* For each receipt, `traverse` calls the provided iterator function https://github.com/onflow/flow-go/blob/82113bba53e963f96dd235ce15cd910dcd73874c/storage/badger/operation/common.go#L345-L346), which re-creates `check`, `create` and `handle`. I don't think this is necessary. These functions can be created _once_ at the beginning of the query and then and reused. 
* Lastly, I think we can avoid the length check. The _current_ implementation essentially guarantees that this check is always satisfied. If the code should be inconsistently modified in the future, some tests should immediately fail, as we will be unable to retrieve previously-stored receipts. If we don't have such tests, lets implement them. This allows us to protect against future inconsistent code modifications at software creation time -- as opposed to a runtime check, which will cost us performance cost on each call.

Putting it all together, I would suggest to use the following implementation
```golang
func receiptIterationFunc(receiptIDs *[]flow.Identifier) func() (checkFunc, createFunc, handleFunc) {
	offset := len(flow.Identifier{}) + 1

	check := func(key []byte) bool {
		var val flow.Identifier
		copy(val[:], key[offset:])
		*receiptIDs = append(*receiptIDs, val)
		return false
	}

	var dummy struct{}
	create := func() interface{} { return &dummy }
	handle := func() error { return nil }

	return func() (checkFunc, createFunc, handleFunc) {
		return check, create, handle
	}
}
``` 
Arguably, this is a little bit hacky. 

#### option 2 favours readability
From my perspective, the complexity in option 1 originates from the fact that we are essentially using the key to store the desired information (not the value). Alternatively, if we want less hacky code with better readability (at the cost of some performance overhead), we could:
1. only use `codeAllBlockReceipts`, `blockID` as prefix and store a _slice_ of receipt IDs
   * the cost here is that we serialize and de-serialize _all_ the receipt IDs every time we add a new receipt ID
   * reads of all receipt IDs for a block are probably equally fast (if not even faster, as we only have to look up one key)
   * overall disk consumption would also be lower, as we still only store one IDs for each receipt , but only the block ID once.  
2. Alternatively, we can store the receipt IDs as value (which is redundant):
   ``` golang
   func IndexExecutionReceipts(blockID, receiptID flow.Identifier) func(*badger.Txn) error {
   	return insert(makePrefix(codeAllBlockReceipts, blockID, receiptID), receiptID)
   }
   ```
   But it is really readable, as it follows the established convention.

Comment (durkmurder) [reply storage/badger/operation/receipts.go:52]: Personally I like:
`insert(makePrefix(codeAllBlockReceipts, blockID, receiptID), receiptID)` the most. But I couldn't just waste memory so did `insert(makePrefix(codeAllBlockReceipts, blockID, receiptID), struct{}{})` since logically it makes sense in the end we are indexing by two keys. I don't really mind and go either way. So someone needs to make a decision :)

File: engine/access/rpc/backend/backend_test.go [hunk 6]
```diff
@@ -602,7 +602,7 @@ func (suite *Suite) TestGetEventsForBlockIDs() {
 		// create receipt mocks that always returns empty
 		receipts := new(storagemock.ExecutionReceipts)
 		receipts.
-			On("ByBlockIDAllExecutionReceipts", mock.Anything).
+			On("ByBlockID", mock.Anything).
 			Return([]*flow.ExecutionReceipt{}, nil).Once()
```
Comment (zhangchiqing) [inline engine/access/rpc/backend/backend_test.go:606]: The mock implements differently than the actual implementation which will return `nil, storage.ErrNotFound`

File: storage/badger/receipts.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (zhangchiqing) [reply storage/badger/receipts.go]: I would suggest to return empty slice instead of `storage.NotFoundError`, because:
1) [access node is expecting to return empty slice here](https://github.com/onflow/flow-go/pull/486/files#diff-ac13d40a77a065ec65a685ed828d76f849a988759b47ce8c2b6f1a8750b7aed6R215-R228), if we return a not found error, it might produce a confusing error message.
1) `ByBlockID` is used by both ExecutionReceipts and MyExecutionReceipts, it's better to keep the behavior consistent by both returning empty slice when receipts are not found by the block ID.

File: storage/badger/my_receipts.go [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+			// check if we are storing same receipt
+			if errors.Is(err, storage.ErrAlreadyExists) {
+				var savedReceiptID flow.Identifier
+				err := operation.LookupOwnExecutionReceipt(blockID, &savedReceiptID)(tx)
+				if err != nil {
+					return err
+				}
+
+				// if we are storing same receipt we shouldn't error
+				if savedReceiptID == receiptID {
```
Comment (zhangchiqing) [inline storage/badger/my_receipts.go:45]: If already exists but doesn't match with the saved receipt ID, I would suggest to include the saved receipt ID in the the error to return , helpful for debugging.

File: storage/badger/my_receipts.go [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module"
+	"github.com/onflow/flow-go/module/metrics"
+	"github.com/onflow/flow-go/storage"
+	"github.com/onflow/flow-go/storage/badger/operation"
+)
+
+// MyExecutionReceipts holds and indexes Execution Receipts.
+// MyExecutionReceipts is implemented as a wrapper around badger.ExecutionReceipts
+// The wrapper ads the ability to "MY execution receipt", from the viewpoint
```
Comment (zhangchiqing) [inline storage/badger/my_receipts.go:18]:
```suggestion
// The wrapper adds the ability to "MY execution receipt", from the viewpoint
```

File: storage/receipts.go [hunk 9]
```diff
@@ -14,24 +18,19 @@ type ExecutionReceipts interface {
...
+	// (from any Execution Node).
+	ByBlockID(blockID flow.Identifier) ([]*flow.ExecutionReceipt, error)
+}
 
-	// IndexByExecutor indexes an execution receipt by block ID and execution ID
-	IndexByExecutor(receipt *flow.ExecutionReceipt) error
+// MyExecutionReceipts holds and indexes Execution Receipts.
+// MyExecutionReceipts is an extension of storage.ExecutionReceipts API.
+// It introduces the notion of "MY execution receipt", from the viewpoint
+// of an individual Execution Node.
```
Comment (zhangchiqing) [inline storage/receipts.go:29]: Good idea and nice refactor 👍  It makes the difference more clear. 
Suggested to update the comments to explain why it's not an `extension` of the `storage.ExecutionReceipts` API
```suggestion
// MyExecutionReceipts reuses the storage.ExecutionReceipts API, but doesn't expose
// them. Instead, it includes the "My" in the method name in order to highlight the notion
// of "MY execution receipt", from the viewpoint of an individual Execution Node.
```

File: cmd/execution/main.go [hunk 10]
```diff
@@ -377,7 +380,11 @@ func main() {
 		}).
 		Component("receipt provider engine", func(node *cmd.FlowNodeBuilder) (module.ReadyDoneAware, error) {
 			retrieve := func(blockID flow.Identifier) (flow.Entity, error) {
```
Comment (AlexHentschel) [inline cmd/execution/main.go:382]: can we replace this and use (?)
```suggestion
				retrieve := myReceipts.MyReceipt
```

there is still some mismatch as `MyReceipt` returns `*flow.ExecutionReceipt`, while `RetrieveFunc` returns an `flow.Entity`.

My suggestion would be to just hand `myReceipts` to the execution Node's `provider.Engine`

File: engine/verification/assigner/indexer.go [hunk 11]
```diff
@@ -27,7 +27,7 @@ func NewExecutionReceiptsIndexer(receipts storage.ExecutionReceipts) *ExecutionR
 // identifier of the executor of receipt.
 func (e *ExecutionReceiptsIndexer) Index(receipts []*flow.ExecutionReceipt) error {
 	for _, receipt := range receipts {
-		err := e.receipts.IndexByExecutor(receipt)
+		err := e.receipts.Store(receipt)
```
Comment (AlexHentschel) [inline engine/verification/assigner/indexer.go:30]: I think the receipt will be stored _and indexed_ as soon as the Consensus follower extends the protocol state with a new (unfinalized) block. I think we might not need the indexer anymore at all, but not sure. 

@zhangchiqing can we remove the indexer?

File: storage/badger/operation/receipts.go (hunk 4 above)
Comment (AlexHentschel) [reply storage/badger/operation/receipts.go:53]: lets waste memory. For a rate of `1 block/s`, 8 Execution nodes, we would wast about 500MB a _year_. I think that is totally acceptable.

File: engine/verification/assigner/indexer.go (hunk 11 above)
Comment (zhangchiqing) [reply engine/verification/assigner/indexer.go:30]: Yes, good catch. We could remove this indexer entirely.

Comment (zhangchiqing) [reply engine/verification/assigner/indexer.go:30]: cc @yhassanzadeh13

File: storage/badger/operation/receipts.go [hunk 12]
```diff
@@ -16,46 +16,43 @@ func RetrieveExecutionReceiptMeta(receiptID flow.Identifier, meta *flow.Executio
...
+	}
+
+	var receiptID flow.Identifier
+	create := func() interface{} {
+		return &receiptID
+	}
+	handle := func() error {
+		*receiptIDs = append(*receiptIDs, receiptID)
+		return nil
+	}
```
Comment (AlexHentschel) [inline storage/badger/operation/receipts.go:54]: ⚠️ I am worried that this will not work. 
We are de-serializing the value into `receiptID` (fixed memory slot) and overwrite the values. I think for this implementation, we need to re-instantiate the `create` and `handle` methods.

Comment (AlexHentschel) [reply storage/badger/operation/receipts.go:54]: can we please add a test for this?

Comment (AlexHentschel) [reply storage/badger/operation/receipts.go:54]: Thanks for extending the test Yuri. 

Please disregard my earlier concern, the code it works. Didn't pay attention to this line, which _copies the value_: https://github.com/onflow/flow-go/blob/b3e8a7de4f46ba2888f3813e4602ed96550d52b3/storage/badger/operation/receipts.go#L52

File: engine/verification/assigner/indexer.go (hunk 11 above)
Comment (yhassanzadeh13) [reply engine/verification/assigner/indexer.go]: @AlexHentschel @zhangchiqing we basically index the receipts in the code snippet above based on their _executor id_. However, I could not find such logic in the implementation of `receipts.Store` anywhere in this branch. The receipts are solely indexed based on their ID as well as their block ID. It introduces two concerns:
(1) The refactored code above does not serve the intended purpose behind indexing a receipt at the verification node. 
(2) Even if we rely on the follower engine indexing the receipts, we still do not get _indexing based on executor_. 

I have also checked this branch and see the `receipts.Store` is no longer invoked upon extending the protocol state. I am not fully knowledgeable in this context, hence may just missing some (significant) details. Just want to make sure that we are not missing a protocol-level _must have_ logic in between by removing or refactoring the indexer in this way.

Comment (zhangchiqing) [reply engine/verification/assigner/indexer.go]: I think the protocol mutate has ensured if a block has been passed all the validation, it will be stored, and [storing the block](https://github.com/onflow/flow-go/blob/yurii/5326-multiple-receipts-for-same-block/state/protocol/badger/mutator.go#L493) will [store the receipt](https://github.com/onflow/flow-go/blob/32ec95c56774f2367984febe70e89ec63fa92b46/storage/badger/payloads.go#L54), as well as indexing the receipt. And this will all be done in one badger transaction. 

it's better than storing the receipt and indexing it later, which is done in two different badger transactions.

>(2) Even if we rely on the follower engine indexing the receipts, we still do not get indexing based on executor.

I don't think we need to index based on executor. The reason verification node needs to query the receipts by executed block ID, is to know who are the executors that produced a certain execution results, so that it can fetch the chunk data pack from those who have executed the block. 

In other words, as long as we can query all receipts by the block ID, then we should be good.

Comment (yhassanzadeh13) [reply engine/verification/assigner/indexer.go]: I see thanks for clarification @zhangchiqing .

**Review by AlexHentschel — APPROVED** [review]
very nice 👏

File: engine/verification/assigner/indexer.go (hunk 11 above)
Comment (AlexHentschel) [reply engine/verification/assigner/indexer.go]: thanks for the detailed comment @zhangchiqing 🏅
